examples_tests/20.Megatexture/main.cpp

```diff
@@ -1,44 +1,378 @@
+	//TODO if (pgtabLoD<LoD) then we're going into mip-tail page! (extra offsets needed)
+	//TODO not sure about correctness of those 3 lines below, seems weird and i dont really understand
+	vec2 tileFractionalCoordinate = virtualUV.xy-vec2(tileCoord)/tilesInLodLevel;
+	tileFractionalCoordinate *= (PAGE_SZ/(PAGE_SZ+TILE_PADDING)); // take border into account
+	physicalUV.xy += tileFractionalCoordinate;
+	return textureGrad(physicalTileTexture,physicalUV,gradients);
+}
 
-	virtual void OnSetConstants(video::IMaterialRendererServices* services, int32_t userData)
-	{
-		services->setShaderConstant(services->getVideoDriver()->getTransform(video::EPTS_PROJ_VIEW_WORLD).pointer(), mvpUniformLocation, mvpUniformType, 1);
-	}
+float lengthSq(in vec2 v)
+{
+  return dot(v,v);
+}
+// textureGrad emulation
+vec4 vTextureGrad(in vec2 virtualUV, mat2 dOriginalUV, in vec2 originalTextureSize)
+{
+  // returns what would have been `textureGrad(originalTexture,gOriginalUV[0],gOriginalUV[1])
+  // https://www.khronos.org/registry/vulkan/specs/1.2-extensions/html/chap15.html#textures-normalized-operations
+  const float kMaxAnisotropy = float(2*TILE_PADDING);
+  const float kMaxAnisoLogOffset = log2(kMaxAnisotropy);
+  // you can use an approx `log2` if you know one
+  float p_x_2_log2 = log2(lengthSq(dOriginalUV[0]*originalTextureSize));
+  float p_y_2_log2 = log2(lengthSq(dOriginalUV[1]*originalTextureSize));
+  bool xIsMajor = p_x_2_log2>p_y_2_log2;
+  float p_min_2_log2 = xIsMajor ? p_y_2_log2:p_x_2_log2;
+  float p_max_2_log2 = xIsMajor ? p_x_2_log2:p_y_2_log2;
+  float LoD = max(p_min_2_log2,p_max_2_log2-kMaxAnisoLogOffset);
+  int LoD_high = int(LoD);
+#ifdef DRIVER_HAS_GRADIENT_SCALING_ISSUES // don't use if you don't have to
+  // normally when sampling an imageview with only 1 miplevel using `textureGrad` the gradient vectors can be scaled arbitrarily because we cannot select any other mip-map
+  // however driver might want to try and get "clever" on us and decide that since `p_min` is huge, it wont bother with anisotropy
+  float anisotropy = min(exp2((p_max_2_log2-p_min_2_log2)*0.5),kMaxAnisotropy);
+  dOriginalUV[0] = normalize(dOriginalUV[0])*(1.0/float(MEGA_TEXTURE_SIZE));
+  dOriginalUV[1] = normalize(dOriginalUV[1])*(1.0/float(MEGA_TEXTURE_SIZE));
+  dOriginalUV[xIsMajor ? 0:1] *= anisotropy;
+#endif
+  return mix(vTextureGrad_helper(virtualUV,LoD_high,dOriginalUV),vTextureGrad_helper(virtualUV,LoD_high+1,dOriginalUV),LoD-float(LoD_high));
+}
+
+vec2 unpackVirtualUV(in uvec2 texData)
+{
+	return unpackUnorm2x16(texData.x);
+}
+vec2 unpackSize(in uvec2 texData)
+{
+	return unpackUnorm2x16(texData.y);
+}
+vec4 textureVT(in uvec2 _texData, in vec2 uv, in mat2 dUV)
+{
+    vec2 scale = unpackSize(_texData);
+    vec2 virtualUV = unpackVirtualUV(_texData);
+    virtualUV += scale*uv*PAGE_SZ_RECIPROCAL;
+    return vTextureGrad(virtualUV, dUV, vec2(textureSize(pgTabTex,0)));
+}
+)";
+constexpr const char* GLSL_BSDF_COS_EVAL_OVERRIDE =
```

Comment:
why override the whole BSDF, if you could only override the texture fetching functions?

Reply:
because GLSL's `texture()` and my `textureVT` differ too much. They have different number of parameters and different types of them. It's very hard/impossible to make a suitable abstraction for texture fetching

Reply:
well the idea is not that you override the GLSL `texture` call, but you override a the MTL shader's `loadKd`, `loadKa` functions etc.

But of course that would require the routines to be overloadable, not just plain `Kd = texture` in the original shader